Add a generic enum-as-number serdes

mjball opened the pull request:
This is particularly useful when you want to serialize a specific field on a
specific type in this manner, without modifying the root object mapper config.

@jhaber

Commit:
Add a generic enum-as-number serdes

This is particularly useful when you want to serialize a specific field on a
specific type in this manner, without modifying the root object mapper config.

mjball commented on src/main/java/com/hubspot/jackson/datatype/protobuf/ProtobufEnumAsNumber.java:
```diff
@@ -0,0 +1,79 @@
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ProtocolMessageEnum;
+
+public final class ProtobufEnumAsNumber {
```
I'm happy to reorganize this into two separate classes if you prefer that style.

mjball commented on src/main/java/com/hubspot/jackson/datatype/protobuf/ProtobufEnumAsNumber.java:
```diff
@@ -0,0 +1,79 @@
+      int intValue = jsonParser.getIntValue();
+      Class<?> enumClass = Preconditions.checkNotNull(javaType, "javaType").getRawClass();
+      try {
+        return (ProtocolMessageEnum) enumClass.getMethod("forNumber", int.class).invoke(null, intValue);
```
Not sure if it's worth memoizing the class &rarr; method mapping, but I'm not super concerned about the performance implications of this implementation in the context I'd like to use this (deserializing from SQL via Rosetta).

jhaber replied:
This could be moved into the `Deserializer` constructor but I don't remember off the top of my head whether contextual deserializers get cached automatically

mjball replied:
I moved it into `createContextual()`. 9d513ec

mjball commented on src/main/java/com/hubspot/jackson/datatype/protobuf/ProtobufEnumAsNumber.java:
```diff
@@ -0,0 +1,79 @@
+      try {
+        return (ProtocolMessageEnum) enumClass.getMethod("forNumber", int.class).invoke(null, intValue);
+      } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+        context.reportWrongTokenException(ProtocolMessageEnum.class, JsonToken.VALUE_NUMBER_INT, wrongTokenMessage(context));
```
I borrowed this pattern from the other deserializer implementations, but I'm not sure if this is really a wrong-token exception.

mjball commented on src/test/java/com/hubspot/jackson/datatype/protobuf/ProtobufEnumAsNumberTest.java:
```diff
@@ -0,0 +1,56 @@
+import com.hubspot.jackson.datatype.protobuf.util.TestProtobuf;
+import com.hubspot.jackson.datatype.protobuf.util.TestProtobuf.Enum;
+
+public class ProtobufEnumAsNumberTest {
```
Happy to add more test cases if you think there are major blind spots.

jhaber commented:
lgtm, might be worth moving the reflective method lookup to the `Deserializer` constructor (assuming that the deserializers get reused by Jackson)

Commit:
Move reflection out of `deserialize()`

https://github.com/HubSpot/jackson-datatype-protobuf/pull/59#discussion_r263135496

mjball commented:
Hm, not sure if there's a way to move the reflective method lookup out of `deserialize()` and memoize its result as a field in the serializer, because `StdDeserializer` is `Serializable` but `java.lang.Method` is not:

```
[INFO] --- findbugs-maven-plugin:3.0.3:check (default) @ jackson-datatype-protobuf ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[INFO] Class com.hubspot.jackson.datatype.protobuf.ProtobufEnumAsNumber$Deserializer defines non-transient non-serializable instance field forNumberMethod [com.hubspot.jackson.datatype.protobuf.ProtobufEnumAsNumber$Deserializer] In ProtobufEnumAsNumber.java SE_BAD_FIELD
[INFO]
```

Commit:
Fix Findbugs SE_BAD_FIELD build failure

Commit:
Suppress more findbugs serialization nonsense

mjball commented:
I got this building again. Feel free to merge and cut the necessary releases whenever, I'm not in a massive hurry to get this feature out.

Commit:
Organize imports
